### ISSUE TITLE
"foo\\" produces incorrect parse error

### ISSUE THREAD

**Issue by kballard** [mawww/kakoune#1049]
If I terminate any double-quoted string with `\\`, Kakoune throws a parse error regarding an unterminated string, as if I had ended it with just `\` (which would then escape the double quote).

Example:

```
echo "foo\\x" # works
echo "foo\\" # parse error
echo "\\" # parse error
```

Similarly, putting `\\` before a `"` escapes it just as putting merely `\` does:

```
echo "foo\\"bar" # this works, but shouldn't
```

As near as I can tell, it's parsing for the end of the string separately from how it parses escapes, and when it finds a `"` it simply looks one character back to see if it's a `\` and if so, treats the quote as escaped, even if that `\` had itself been escaped.

**Comment by kballard** [comment]
It occurs to me that with this fixed, there's one subtlety, which is how is `"%"foo""` parsed? The two options are:

1. Terminate the string literal at the 2nd `"`, leaving you with a string that just contains `%`, which would then presumably throw an error just as it does today if you use `"%"`
2. Parse the 2nd `"` as the start delimiter of the `%{`-expansion, which makes the whole thing parse as one string with an embedded expansion.

I'd lean towards the first simply because, with either approach, `"%"` produces an error, but the error produced from approach 1 makes more sense than complaining about an unterminated string (because it tells the user that the `%` is trying to introduce a string expansion, rather than confusing them by saying there's no string terminator when it looks like there is one).

**Comment by mawww** [comment]
Regarding "%"foo"", it should do the first option, we parse the content on the "..." later on, so the existence of expands in the string content has no impact on how we parse it, we just look for the first non escaped closer.

The other behaviour seems to be an error, I'll take a stab at fixing that when I get a bit of time.

**Comment by kballard** [comment]
I realize that's how you're parsing it today, but I'm saying that behavior isn't very useful, because it makes an unescaped `"` inside a nested escape completely useless, and therefore the requirement to escape nested `"`'s is unnecessary and annoying. The only reason why I'm not saying more strongly that this should be changed is because I'm not sure how much existing code would be affected by this change (probably not much). But if you're ok with the minor backwards-incompatible fix, then I think it really should change.

**Comment by kballard** [comment]
Ignore my previous comment (now deleted), I confused this issue with my other one about quote parsing.

**Comment by mawww** [comment]
Argh, I took a look, its pretty easy to fix, but it brings us into escaping hell, as `\\` now needs to be written `\\\\` in `"..."` and `'...'` strings. Thankfully we have % strings that do not support escaping at all.

**Comment by danr** [comment]
I'm getting bitten by this over at https://github.com/danr/libkak where I want to send a `%sh` in a `hook` in an `eval` and so on... 

Maybe we can repurpose the little used `%'...'` and `%"..."` for this behaviour if you want to keep `'` and `"` as they are?

**Comment by mawww** [comment]
I'd really like to fix this, but finding a nice solution is hard. The current behaviour is to just consider `\"` as meaning `"` in quoted strings, and leave every other `\` as is. Unfortunately that means we cannot end such a string with a `\`.

The obvious solution is to make `\` escapable as well, but that leads us into escaping hell, as it means all the currently used `\` appearing outside of `%{...}` strings in `.kak` files need to be escaped. That means in particular every use of \ in regexes, so all hooks, highlighters, autoindent <kbd>s</kbd>, <kbd>\<a-k></kbd>, <kbd>/</kbd>... That would not only require to fix all existing scripts, but would also make much of them less readable.

An alternate solution would be to follow the shell on parsing `'...'` strings: no escaping at all in single quote strings, and `'` cannot appear in single quote strings. This has the advantage that single quote strings stay very similar to how they work at the moment: We only remove the ability to put a single quote inside them.

That still mean we would need to go through all scripts and fix them, but the fix is far more readable: wrap regex inside single quote strings (if they dont contain a `'`), or wrap them inside a `%{...}` string with a well chosen delimiter.

Another alternative is to change the way we parse non quoted words, by always treating them literally. That means a non quoted word cannot start with '%' and cannot contain whitespaces. This would avoid changing as much of the scripts, as this is close the the common practice, I dont remember seeing many whitespaces escaped in existing scripts.

My main concern with that second approach is consistency, it means we have very different quoting behaviours for various kind of strings.

What are your thought on this ?

**Comment by Screwtapello** [comment]
This is a *little* off-topic, but I got confused the other day when I tried to run:

    set-option -add global ui_options gtk_font="DejaVu Sans Mono 10"

...and got a weird error about the font `"DejaVu` not existing. I was expecting shell-style quoting where quotes can be anywhere and just protect their contents.

As such, following shell's lead with single-quoted strings is appealing to me; since Kakoune scripting already depends so heavily on the shell, we can pretty safely assume Kakoune scripters understand how shell-quoting works (and if not, they'll be learning about it very quickly anyway). Following the shell model, I imagine double-quoted strings would need the double-quote character to be backslash-escaped (of course), as well as backslash itself (as described above) and also `%` to avoid unwanted expansions (much as `$` can be escaped in shell).
